lace/archive/gadget_archive.py

```diff
@@ -270,22 +270,28 @@ def _get_nz_linP(self, sim_label):
 
         """
 
-        _, sim_name_param, tag_param = self._sim2file_name(sim_label)
-        pair_dir = self.fulldir_param + "/" + sim_name_param
-
         # read zs values and compute/read linP_zs
-        if (self.update_kp == False) & (sim_label in self.list_sim_cube):
-            pair_json = pair_dir + "/" + tag_param
+        if self.update_kp == False:
+            emu_cosmo_all = np.load(
+                self.fulldir + "mpg_emu_cosmo.npy", allow_pickle=True
+            )
+            for ii in range(len(emu_cosmo_all)):
+                if emu_cosmo_all[ii]["sim"] == sim_label:
+                    if emu_cosmo_all[ii]["emu_cosmo"]["kp_Mpc"] != self.kp_Mpc:
+                        self.update_kp = True
+                    else:
+                        cosmo = emu_cosmo_all[ii]["cosmo"]
+                        emu_cosmo = emu_cosmo_all[ii]["emu_cosmo"]
+                    break
+
+        if self.update_kp == True:
+            _, sim_name_param, tag_param = self._sim2file_name(sim_label)
+            pair_dir = self.fulldir_param + "/" + sim_name_param
 
-            with open(pair_json) as json_file:
-                pair_data = json.load(json_file)
-            zs = pair_data["zs"]
-            linP_zs = pair_data["linP_zs"]
-        else:
             # read gadget file
             gadget_fname = pair_dir + "/sim_plus/paramfile.gadget"
-            sim_config = read_gadget.read_gadget_paramfile(gadget_fname)
-            zs = read_gadget.snapshot_redshifts(sim_config)
+            cosmo = read_gadget.read_gadget_paramfile(gadget_fname)
```

Comment:
I would not call this "cosmo", this is the configuration file of the Gadget runs, that contains cosmo information but also many other things (in the next line you use it to get the redshifts).